src/github.com/getlantern/flashlight/ui/ws.go

```diff
@@ -0,0 +1,153 @@
+package ui
+
+import (
+	"io"
+	"net/http"
+	"path"
+	"sync"
+
+	"github.com/gorilla/websocket"
+)
+
+const (
+	// Determines the chunking size of messages used by gorilla
+	MaxMessageSize = 1024
+)
+
+var (
+	upgrader = &websocket.Upgrader{ReadBufferSize: 1024, WriteBufferSize: MaxMessageSize}
+)
+
+// UIChannel represents a data channel to/from the UI. UIChannel will have one
+// underlying websocket connection for each connected browser window. All
+// messages from any browser window are available via In and all messages sent
+// to Out will be published to all browser windows.
+type UIChannel struct {
+	URL string
+	In  <-chan []byte
+	Out chan<- []byte
+
+	in     chan []byte
+	out    chan []byte
+	nextId int
+	conns  map[int]*wsconn
+	m      sync.Mutex
+}
+
+type ConnectFunc func(write func([]byte) error) error
+
+// NewChannel establishes a new channel to the UI at the given path. When the UI
+// connects to this path, we will establish a websocket to the UI to carry
+// messages for this UIChannel. The given onConnect function is called anytime
+// that the UI connects.
+func NewChannel(p string, onConnect ConnectFunc) *UIChannel {
+	c := newUIChannel(path.Join(uiaddr, p))
+
+	r.HandleFunc(p, func(resp http.ResponseWriter, req *http.Request) {
+		log.Tracef("Got connection to %v", c.URL)
+		var err error
+
+		if req.Method != "GET" {
+			http.Error(resp, "Method not allowed", 405)
+			return
+		}
+		// Upgrade with a HTTP request returns a websocket connection
+		ws, err := upgrader.Upgrade(resp, req, nil)
+		if err != nil {
+			log.Errorf("Unable to upgrade %v to websocket: %v", p, err)
+			return
+		}
+
+		log.Tracef("Upgraded to websocket at %v", c.URL)
+		c.m.Lock()
+		if onConnect != nil {
+			err = onConnect(func(b []byte) error {
+				log.Tracef("Writing initial message: %q", b)
+				return ws.WriteMessage(websocket.TextMessage, b)
+			})
+			if err != nil {
+				log.Errorf("Error processing onConnect, disconnecting websocket: %v", err)
+				ws.Close()
+				c.m.Unlock()
+				return
+			}
+		}
+		c.nextId += 1
+		conn := &wsconn{
+			id: c.nextId,
+			c:  c,
+			ws: ws,
+		}
+		c.conns[conn.id] = conn
+		c.m.Unlock()
+		log.Tracef("About to read from connection to %v", c.URL)
+		conn.read()
+	})
+
+	return c
+}
+
+func newUIChannel(url string) *UIChannel {
+	in := make(chan []byte, 100)
+	out := make(chan []byte)
```

Comment:
I noticed that the out channel is not buffered (`newUIChannel()`), so a goroutine that writes to it would block until the message is received on the client by WriteMessage. Since we're already thinking on a situation when two or more browsers are opened it would probably be better to have a small buffer here.